### ISSUE TITLE
performance degredation using cudaMallocManaged (Unified Memory) instead of cudaMalloc

### ISSUE THREAD

**Issue by mcginty** [NVIDIA/NvPipe#45]
When using cudaMallocManaged, the performance in the `memory` example is exactly equal to simply using host memory.

Forgive me if I'm misunderstanding, but it seems like while there's some performance difference between using `cudaMalloc` and `cudaMallocManaged`, it shouldn't be this extreme?

This was tested on a GTX 1080.

### with `cudaMalloc()`
```
NvPipe example application: Comparison of using host/device memory.

Resolution: 3840 x 2160
Codec: HEVC
Bitrate: 32 Mbps @ 60 Hz
Resolution: 3840 x 2160

--- Encode from host memory / Decode to host memory ---
Frame | Encode (ms) | Decode (ms) | Size (KB)
    0 |        74.2 |        47.5 |     47.4
    1 |        26.2 |        17.0 |     14.2
    2 |        26.1 |        17.1 |     10.4
    3 |        26.0 |        16.9 |      8.2
    4 |        26.1 |        17.1 |      7.5
    5 |        26.2 |        17.1 |      6.1
...

--- Encode from device memory / Decode to device memory ---
Frame | Encode (ms) | Decode (ms) | Size (KB)
    0 |        59.9 |        34.3 |     47.4
    1 |        12.5 |         5.1 |     14.2
    2 |        12.2 |         5.1 |     10.4
    3 |        12.3 |         4.9 |      8.2
    4 |        12.3 |         5.0 |      7.5
    5 |        12.3 |         5.0 |      6.1
...
```

### with `cudaMallocManaged(..., cudaMemAttachGlobal)`
```
--- Encode from device memory / Decode to device memory ---
Frame | Encode (ms) | Decode (ms) | Size (KB)
    0 |        93.6 |        47.6 |     47.4
    1 |        26.2 |        16.8 |     14.2
    2 |        26.2 |        17.3 |     10.4
    3 |        26.3 |        16.9 |      8.2
    4 |        26.4 |        17.3 |      7.5
    5 |        26.5 |        17.0 |      6.1
    6 |        26.4 |        17.2 |      6.1
```

**Comment by tbiedert** [comment]
Hi, thanks for reaching out!

It's possible that `cudaMallocManaged` returns host memory. Please check the [documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gd228014f19cc0975ebe3e0dd2af6dd1b).

You could try getting some info about the returned pointer via [cudaPointerGetAttributes ](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390), or alternatively using nvvp.

**Comment by mcginty** [comment]
@tbiedert thanks so much for the response! nvvp is awesome.

nvvp seems to confirm that `cudaMallocManaged` is allocating device memory, but or some reason, the `memory` example spends  50x as much GPU time on DtoD. The avg. DtoD throughput in the managed case is 3.4GB/s, but is 119GB/s in the normal `cudaMalloc` case. Puzzling, and I don't think I know enough yet about the different types of device allocations to have a good guess as to why this behavior exists.

Here are the NVVP files, but it's easy enough to reproduce.
[profiles.zip](https://github.com/NVIDIA/NvPipe/files/2735265/profiles.zip)

**Comment by tbiedert** [comment]
Based on the profiles I suppose you are on Linux, right?  Which CUDA version are you running?

On Windows, I could not reproduce the problem. However, I discovered a bug in NvPipe's internal host/device pointer detection, which can cause problems with unified memory.

Please try replacing the `bool isDevicePointer` function in `NvPipe.cu` with the following code:
```c++
inline bool isDevicePointer(const void* ptr)
{
    struct cudaPointerAttributes attr;
    const cudaError_t perr = cudaPointerGetAttributes(&attr, ptr);

#if (CUDA_VERSION >= 10000)
    return (perr == cudaSuccess) && (attr.type != cudaMemoryTypeHost);
#else
    return (perr == cudaSuccess) && (attr.memoryType != cudaMemoryTypeHost);
#endif
}
```

Thanks!

**Comment by mcginty** [comment]
That fixed it!! 🎉🎉 Great catch.

Oh, sorry for not specifying. I'm on Arch Linux with the latest cuda (10.0.130-2) package in the community repository.

**Comment by tbiedert** [comment]
Thanks for testing! I've pushed the fix in 4f410170332503eccd59d9e3c29ae600353ca556.
